Allow .parse to load from string

kmcphillips opened the pull request:
I'm doing some work of loading weights from files, and I want to be able to have values in an input file:

```
weight: 10.5 kg
```

And then create a weight object out of it:

```
Measured::Weight.parse("10.5 kg")
```

It's way slower than taking in the number and the symbol, but we'd be parsing those strings from the file first anyway.

kmcphillips commented on lib/measured/measurable.rb:
```diff
@@ -1,6 +1,8 @@
 class Measured::Measurable < Numeric
   include Measured::Arithmetic
 
+  PARSE_REGEX = /\A\s*([0-9.]+)\s*([a-zA-Z_-]+)\s*\Z/
```
Accept only one line, any blank characters, then any digits with an optional dot, then optional blank characters, than any letters numbers dashes underscores, then any blank characters.

thegedge replied:
`Unit` does [this](https://github.com/Shopify/measured/blob/2a05bfc90ea95a29fc11cd693b20c9e84eec2be7/lib/measured/unit.rb#L66-L71), for conversion values when building a unit system, so we should probably share the parsing code between them.

Two cases missing atm: negative numbers (e.g., -1°C) and unit names with spaces (e.g., nautical mile).

jonathankwok replied:
So the fact that we need a review comment means we should do something about the regex clarity. Free-spacing regex is pretty fantastic, by using `/x` which ignores whitespaces.

```ruby
# before
/\A\s*([0-9.]+)\s*([a-zA-Z_-]+)\s*\Z/

# after
/
  \A                # start of line
  \s*               # any number of blank characters
  ([0-9.]+)         # one or more digits or decimals
  \s*               # any number of blank characters
  ([a-zA-Z_-]+)     # one or more letters, dashes, or underscores
  \s*               # any number of blank characters
  \Z                # end of line
/x
```

jonathankwok replied:
Also in doing that I see that we accept multiple dots. Does that make sense for measured? Numbers/quantities can't have multiple dots.

```ruby
irb(main):001:0> pattern = /
irb(main):002:0/   \A                # start of line
irb(main):003:0/   \s*               # any number of blank characters
irb(main):004:0/   ([0-9.]+)         # one or more digits or decimals
irb(main):005:0/   \s*               # any number of blank characters
irb(main):006:0/   ([a-zA-Z_-]+)     # one or more letters, dashes, or underscores
irb(main):007:0/   \s*               # any number of blank characters
irb(main):008:0/   \Z                # end of line
irb(main):009:0/ /x
=> /
  \A                # start of line
  \s*               # any number of blank characters
  ([0-9.]+)         # one or more digits or decimals
  \s*               # any number of blank characters
  ([a-zA-Z_-]+)     # one or more letters, dashes, or underscores
  \s*               # any number of blank characters
  \Z                # end of line
/x
irb(main):010:0> pattern =~ "1.21.32 gigawatts"
=> 0
```

jonathankwok replied:
Finally, on that topic - we should consider commas and periods inter-changable. [A non-trivial amount of countries in the world represent the idea of `1.21` as `1,21`.](http://brilliantmaps.com/decimals/)

kmcphillips commented:
Fantastic answers!

Splitting out the regex for readability is a wonderful idea. I'll do that, and deal with negatives and single period. I'll also test those cases.

I'll see about refactoring the `Unit` class to share this logic. I'm sure we can use a single parser, or at least a single approach. I really don't want to just do `.strip.split(" ", 2)` because I want to support `12cm` as that's easy to parse but has no spaces.

jonathankwok commented on lib/measured/measurable.rb:
```diff
@@ -1,6 +1,26 @@
 class Measured::Measurable < Numeric
   include Measured::Arithmetic
 
+  PARSE_REGEX = /
```
![](http://i0.kym-cdn.com/photos/images/newsfeed/001/074/167/3f1.gif)

jonathankwok commented on lib/measured/measurable.rb:
```diff
@@ -1,6 +1,26 @@
+      -?       # number can be negative
+      \d+      # must have some digits
+      (?:      # do not capture
+        \.     # period to split fractional part
```
We should permit the use of commas in place of periods. (https://github.com/Shopify/measured/pull/81#discussion_r102552174)

thegedge replied:
I think commas are pervasive enough to support, but a larger question to ask is how far do we go with localization here. Can take that to a :octocat: issue.

thegedge commented on README.md:
````diff
@@ -56,6 +56,12 @@ rescue Measured::UnitError
+Parse from string without having to split out the value and unit first:
+
+```ruby
+Measured::Weight.parse("123 grams")
````
nit: to be aligned with the other comments, show the `#inspect` after. Maybe also show that you don't need a space.

kmcphillips replied:
We don't show the result for all operations. And this is just showing the parser syntax.

kmcphillips replied:
But I think you're right anyway. 😄 

Added more detail.

thegedge replied:
Up to you :)

thegedge commented on lib/measured/measurable.rb:
```diff
@@ -1,6 +1,26 @@
+    )
+    \s*        # optionally any whitespace
+    \Z         # end of unit
+  /x
```
We can iterate on it, but some other things I'd love to be able to support:

- `1/3 grams`
- `5e-10 cm`
- `.5 metres`

thegedge commented on lib/measured/measurable.rb:
```diff
@@ -1,6 +1,26 @@
+      +?       # capture it lazily to not take trailing whitespace
+    )
+    \s*        # optionally any whitespace
+    \Z         # end of unit
```
nit: "end of input"

thegedge commented on test/measurable_test.rb:
```diff
@@ -94,7 +94,7 @@ class Measured::MeasurableTest < ActiveSupport::TestCase
   test ".unit_names_with_aliases returns all units" do
     assert_equal(
-      %w(arcane fire fireball fireballs ice magic_missile magic_missiles ultima),
+      %w(arcane fire fireball fireballs ice magic\ missile magic_missile magic_missiles ultima),
```
😭 

But 👌  for having this case tested 😄

kmcphillips commented:
Ok, reping.

Since we parse in both the unit definition and now on instantiating a measurement, I extracted a `Measured::Parser` class. It's nice because it now encapsulates the regex. Split it up as suggested, with comments. Way built out what's in that regex. Added a ton of edge cases. Shared the parsing.

Some tests are duplicated, but I'm ok with that. Had to add a case for a unit with a space in it.

If it's too much I can split the PR up. But I think it's easy enough to follow.

kmcphillips commented:
Oh and @jonathankwok I went back and forth on the `,` parsing and decided _not_ to include it. I would be doing a `value.gsub(",", ".")` since the numeric parsing in ruby truncates after the comma. It doesn't align with the language, so we won't do it here.

kmcphillips commented on lib/measured/unit.rb:
```diff
@@ -64,9 +64,14 @@ def conversion_string
   def parse_value(tokens)
-    tokens = tokens.split(" ", 2) if tokens.is_a?(String)
-
-    raise Measured::UnitError, "Cannot parse 'number unit' or [number, unit] formatted tokens from #{tokens}." unless tokens.size == 2
```
This is all now much more robust. Error messaging is cleaner, and it allows units to be declared with more edge cases in mind.

jonathankwok commented on lib/measured/parser.rb:
```diff
@@ -0,0 +1,37 @@
+    )
+    \s*        # optionally any space between number and unit
+    (          # capture the unit
+      [a-zA-Z] #
```
This is because we want the unit to start with a letter, right? Maybe a comment to suggest as much?

jonathankwok commented on lib/measured/parser.rb:
```diff
@@ -0,0 +1,37 @@
+    (          # capture the unit
+      [a-zA-Z] #
+      [\w-]*   # any word characters or dashes
+      (?:
```
Can you add comments to explain what this non-capturing group does?

kmcphillips commented:
@jonathankwok much clearer regex comments now

jonathankwok commented on lib/measured/parser.rb:
```diff
@@ -0,0 +1,37 @@
+    )
+    \s*        # optionally any space between number and unit
+    (          # capture the unit
+      [a-zA-Z] # unit maust start with a letter
```
`maust`

jonathankwok commented on lib/measured/parser.rb:
```diff
@@ -0,0 +1,37 @@
+      [a-zA-Z] # unit maust start with a letter
+      [\w-]*   # any word characters or dashes
+      (?:      # non capturing group
+        \s+    # there can be a space in the unit
```
`can`, or `must`? 

`+` means `must`, since it's one or more, `*` means `can` because it's zero or more.

kmcphillips replied:
Yeah. There _can_ be a space, but then after the space there must be another thing. But this whole non-capturing group is optional. I can make that clearer.
